### PR TITLE
Adding support for running tests unpackaged 

### PR DESCRIPTION
**Pull request by loneursid**
Adds support to the for running unpackaged Push Notification tests.

**Usage Example:**
Currently, the code that launch a ChannelRequest unit test looks like this:
````cpp
        TEST_METHOD(ChannelRequestUsingRemoteId)
        {
            RunTest(L"ChannelRequestUsingRemoteId", channelTestWaitTime());
        }
````

With this PR, it will also be possible to run the same unit test for an unpackaged scenario. It will look like this:
````cpp
        TEST_METHOD(ChannelRequestUsingRemoteId_Unpackaged)
        {
            RunTestUnpackaged(L"ChannelRequestUsingRemoteId", channelTestWaitTime());
        }
````

### CODE REVIEW

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 1]
```diff
@@ -4,6 +4,7 @@
 #include <sstream>
 #include <wil/win32_helpers.h>
 #include <winrt/Windows.ApplicationModel.Background.h> // we need this for BackgroundTask APIs
+#include "../../../dev/Common/AppModel.Identity.h"
```
[the remark this answers is not in the document]
Comment (loneursid) [reply test/TestApps/PushNotificationsTestApp/main.cpp]: Done

File: test/PushNotificationTests/APITests.cpp [hunk 2]
```diff
@@ -116,6 +141,26 @@ namespace Test::PushNotifications
             VERIFY_ARE_EQUAL(exitCode, 0);
         }
 
+        const std::wstring GetDeploymentDir()
+        {
+            WEX::Common::String deploymentDir;
+            WEX::TestExecution::RuntimeParameters::TryGetValue(L"TestDeploymentDir", deploymentDir);
+            deploymentDir.Append(L"..\\PushNotificationsTestApp");
```
[the remark this answers is not in the document]
Comment (loneursid) [reply test/PushNotificationTests/APITests.cpp]: Done

File: test/PushNotificationTests/APITests.cpp [hunk 3]
```diff
@@ -58,6 +58,7 @@ namespace Test::PushNotifications
             try
             {
                 TP::AddPackage_WindowsAppSDKFramework(); // Installs WASfwk
+                TP::AddPackage_DynamicDependencyLifetimeManager(); // Required when running the test app unpackaged.
```
Comment (DrusTheAxe) [inline test/PushNotificationTests/APITests.cpp:61]: Gonna need the main package too (that has the DataStore)
If you install WASfwk you should always install the trifecta - WASfwk, WASmain, WASddlm

File: test/PushNotificationTests/APITests.cpp [hunk 4]
```diff
@@ -116,6 +141,25 @@ namespace Test::PushNotifications
...
         }
 
+        const std::wstring GetDeploymentDir()
+        {
+            WEX::Common::String deploymentDir;
+            WEX::TestExecution::RuntimeParameters::TryGetValue(L"TestDeploymentDir", deploymentDir);
+            return reinterpret_cast<PCWSTR>(deploymentDir.GetBuffer());
+        }
+
+        void RunTestUnpackaged(const PCWSTR& testName, const int& waitTime)
```
Comment (DrusTheAxe) [inline test/PushNotificationTests/APITests.cpp:151]: You can do this via TAEF. Worst case define 2 tests, one with RunAs:UWP and one without.

No need to roll your own test process launcher

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 5]
```diff
@@ -18,6 +19,11 @@ winrt::guid remoteId1(L"a2e4a323-b518-4799-9e80-0b37aeb0d225"); // Generated fro
 winrt::guid remoteId2(L"CA1A4AB2-AC1D-4EFC-A132-E5A191CA285A"); // Dummy guid from visual studio guid tool generator
 
 PushNotificationRegistrationToken g_appToken = nullptr;
+const wchar_t* g_bootStrapDllName = L"Microsoft.WindowsAppSDK.Bootstrap.dll";
+
+typedef HRESULT(*BootStrapTestInit)(PCWSTR prefix, PCWSTR publisherId);
+typedef HRESULT(*BootStrapInit)(const UINT32 majorMinorVersion, PCWSTR versionTag, const PACKAGE_VERSION minVersion);
+typedef void (*BootStrapShutdown)();
```
Comment (DrusTheAxe) [inline test/TestApps/PushNotificationsTestApp/main.cpp:26]: Why?

The test\inc\ helpers know about the solution's build output layout and knows how to find the bootstrap .dll (and in the real world the WindowsAppSDK NuGet includes the dll and VS knows to copy it to dependents build output directory)

Use the test helpers in test\inc to call the bootstrap DLL and mark the Bootstrap lib as DelayLoad in your .vcxproj so your exe can load with it, then the test helpers make it accessible and voila.

Much simpler than tests replicating all this code...

**Review by DrusTheAxe — CHANGES_REQUESTED** [review]

File: test/PushNotificationTests/PushNotificationTests.vcxproj.filters [hunk 6]
```diff
@@ -13,6 +13,9 @@
       <UniqueIdentifier>{67DA6AB6-F800-4c08-8B7A-83BB121AAD01}</UniqueIdentifier>
       <Extensions>rc;ico;cur;bmp;dlg;rc2;rct;bin;rgs;gif;jpg;jpeg;jpe;resx;tiff;tif;png;wav;mfcribbon-ms</Extensions>
     </Filter>
+    <Filter Include="CopyFiles">
+      <UniqueIdentifier>{a61a0e17-91e0-4d34-9aaa-a6b871593591}</UniqueIdentifier>
```
Comment (danielayala94) [inline test/PushNotificationTests/PushNotificationTests.vcxproj.filters:17]: What is this for?

File: test/TestApps/PushNotificationsTestApp/PushNotificationsTestApp.vcxproj [hunk 7]
```diff
@@ -255,6 +255,7 @@
   <ImportGroup Label="ExtensionTargets">
     <Import Project="..\..\..\packages\Microsoft.Windows.CppWinRT.2.0.210505.3\build\native\Microsoft.Windows.CppWinRT.targets" Condition="Exists('..\..\..\packages\Microsoft.Windows.CppWinRT.2.0.210505.3\build\native\Microsoft.Windows.CppWinRT.targets')" />
     <Import Project="..\..\..\packages\Microsoft.Windows.ImplementationLibrary.1.0.210204.1\build\native\Microsoft.Windows.ImplementationLibrary.targets" Condition="Exists('..\..\..\packages\Microsoft.Windows.ImplementationLibrary.1.0.210204.1\build\native\Microsoft.Windows.ImplementationLibrary.targets')" />
+    <Import Project="..\..\..\packages\Microsoft.Taef.10.58.210222006-develop\build\Microsoft.Taef.targets" Condition="Exists('..\..\..\packages\Microsoft.Taef.10.58.210222006-develop\build\Microsoft.Taef.targets')" />
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/PushNotificationsTestApp.vcxproj:258]: What is the purpose of importing TAEF to the test app exe?

File: test/PushNotificationTests/PushNotificationTests.vcxproj.filters (hunk 6 above)
Comment (loneursid) [reply test/PushNotificationTests/PushNotificationTests.vcxproj.filters:17]: CopyFiles fiter in the solution. Just to group the extra files that needs to be copied to the output dir together

File: test/TestApps/PushNotificationsTestApp/PushNotificationsTestApp.vcxproj (hunk 7 above)
Comment (loneursid) [reply test/TestApps/PushNotificationsTestApp/PushNotificationsTestApp.vcxproj:258]: Dependencies on the test namespace, like Test::AppModel::IsPackagedProcess(); that we're picking up following @DrusTheAxe comments.

File: test/PushNotificationTests/APITests.cpp (hunk 3 above)
Comment (loneursid) [reply test/PushNotificationTests/APITests.cpp]: Done

File: test/TestApps/PushNotificationsTestApp/main.cpp (hunk 5 above)
Comment (loneursid) [reply test/TestApps/PushNotificationsTestApp/main.cpp]: Done

**Review by DrusTheAxe — APPROVED** [review]

**Pull request merged**